keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -183,70 +184,78 @@ class ApplicationService(
         val artifactSummariesInEnvironments = mutableSetOf<ArtifactSummaryInEnvironment>()
 
         environmentSummaries.forEach { environmentSummary ->
           val environment = deliveryConfig.environments.find { it.name == environmentSummary.name }!!
-          environmentSummary.getArtifactPromotionStatus(artifact, version)?.let { status ->
-            var artEnvSummary = when (status) {
-              PENDING -> ArtifactSummaryInEnvironment(
-                environment = environmentSummary.name,
-                version = version,
-                state = status.name.toLowerCase()
-              )
-              SKIPPED -> {
-                // some environments contain relevant info for skipped artifacts, so
-                // try and find that summary before defaulting to less information
-                val potentialSummary = repository.getArtifactSummaryInEnvironment(
-                  deliveryConfig = deliveryConfig,
-                  environmentName = environmentSummary.name,
-                  artifactReference = artifact.reference,
-                  version = version
-                )
-                if (potentialSummary == null || potentialSummary.state == "pending") {
-                  ArtifactSummaryInEnvironment(
-                    environment = environmentSummary.name,
-                    version = version,
-                    state = status.name.toLowerCase()
+          environmentSummary.getArtifactPromotionStatus(artifact, version)
+            ?.let { status ->
+              buildArtifactSummaryInEnvironment(deliveryConfig, environment.name, artifact, version, status)
+                ?.also {
+                  artifactSummariesInEnvironments.add(
+                    it.addStatefulConstraintSummaries(deliveryConfig, environment, version)
+                      .addStatelessConstraintSummaries(deliveryConfig, environment, version, artifact)
                   )
-                } else {
-                  potentialSummary
                 }
-              }
-              else -> repository.getArtifactSummaryInEnvironment(
-                deliveryConfig = deliveryConfig,
-                environmentName = environmentSummary.name,
-                artifactReference = artifact.reference,
-                version = version
-              )
             }
-            if (artEnvSummary != null) {
-              artEnvSummary = addStatefulConstraintSummaries(artEnvSummary, deliveryConfig, environment, version)
-              artEnvSummary = addStatelessConstraintSummaries(artEnvSummary, deliveryConfig, environment, version, artifact)
-              artifactSummariesInEnvironments.add(artEnvSummary)
-            }
-          }
         }
-        return@map versionToSummary(artifact, version, artifactSummariesInEnvironments.toSet())
+
+        buildArtifactVersionSummary(artifact, version, artifactSummariesInEnvironments)
       }
-      return@map ArtifactSummary(
```

Comment:
I personally think the `return@map` makes it more clear, because I dislike having the last line returned by default for anything more than ~2 lines. It just makes it easier to scan and see what's going on

Reply:
The compiler was issuing warnings for both statements because there are two `.map` calls. That's why I decided to refactor so that the loops are much more concise and fit in a single screen. Hope that's OK.